✨ v3: 마지막 리팩토링

healtheefart opened the pull request:
## 💻 작업 내역

- 단축된 URL에 대한 요청 수(cnt) 정보저장
- 두 가지 Exception에 대한 처리 추가
  - `InvalidArgumentException` : @Validated으로부터 던져진 exception
  - `NotFoundResourceException` : 단축된 적 없는 shorten-url에 대한 redirect가 일어났을 때 던져진 exception
  - `index.html` : error 발생 시 exception에 대한 message 출력
- BASE62Utils에 @Value 어노테이션 수정

## 🔎 PR 특이 사항

- @Validated 어노테이션에 의해 던져지는 exception이 `ConstraintViolationException` 라고 했는데 `MethodArgumentNotValidException`로 발생해서 일단 후자로 exception을 처리하였지만 왜 후자의 exception으로 throw하는지 이유를 모르겠습니다...
- 이전 pr까지는 @Value에 의해 BASE62Utils의 static field 들이 yml 값들로 잘 주입이 됐는데 이번 pr을 진행하면서 갑자기 값 인식을 못하기에 찾아보니 [static한 field에 대해 @Value를 사용하려면 field자체가 아닌 un-static한 setter에 @Value를 붙여서 사용해야한다](https://thingsthis.tistory.com/326)고 하여 그렇게 변경하였습니다. 하지만 그 전엔 왜 문제가 없었는지는...

## 🛠️ refactoring list

- @Validated의 `grouping 기능`을 사용하고자 했는데 현재 서비스에서는 validate할 값이 url밖에 존재하지 않아 사용하지 않았습니다. 다음에 기회가 된다면 해보는 것도 좋을 것 같습니다.
- 현재 url id에 대한 `encoded 값`이 7자가 되지 않는데 encoded값이 7자 이상이 되도록 단축 알고리즘을 수정해보는 것도 좋을 것 같습니다.
- url data에 대한 `만료기능`을 주는 것도 좋을 것 같습니다.

healtheefart committed:
init project

healtheefart committed:
Merge pull request #3 from healtheefart/healtheefart/develop

init project

healtheefart committed:
Revert "init project"

healtheefart committed:
Merge pull request #4 from study-for-Coding-Mission/revert-3-healtheefart/develop

Revert 3 healtheefart/develop

healtheefart committed:
Merge pull request #1 from study-for-Coding-Mission/healtheefart/develop

revert : set project into empty project

healtheefart committed:
re-init project

healtheefart committed:
chore: change properties into yml

healtheefart committed:
chore: added .gitignore

healtheefart committed:
chore: added dependency of mysql, thymeleaf, lombok, jpa

healtheefart committed:
Feat : created BASE62Utils
- (encode) : url의 id를 encode한 값 return
- (decode) : encode한 값을 decode한 url의 id return

healtheefart committed:
Feat : created controller
- (HomeController) : 홈화면, redirect
- (UrlShortenerController) : url 단축

healtheefart committed:
Feat : created UrlShortenerService

healtheefart committed:
Feat : created UrlShortenerRepository

healtheefart committed:
Feat : created ShortenUrl(entity), ShortenUrlRequest(dto), ShortenUrlResponse(dto)

healtheefart committed:
Feat : created home.html(thymeleaf page), style.css

healtheefart committed:
refactor : text에서 JSON형식으로 parse한 응답값으로 반환하도록 수정

healtheefart committed:
refactor : getOriginUrl 중복된 코드 수정

healtheefart committed:
chore : added log4j, springboot validation dependency

healtheefart committed:
refactor : WebConfigurer를 통한 index.html 뷰 반환
- index.html : renamed from home.html
- RedirectController.java
  - renamed from HomeController.java
  - removed getHome()
- ShortenUrlResponse : removed originUrl field

healtheefart committed:
naming & refactor : ShortenUrl.java -> Url.java로 수정
- ShortenUrlRequest와 네이밍을 구별하기 위해 엔티티를 Url.java로 네이밍 수정
- Url Id를 BigInteger로, 시작Value를 100000으로 주어 encode값을 일정 길이 이상이 되게끔 수정

healtheefart committed:
naming & refactor : ShortenUrlDto -> UrlShortenerDto로 수정
- ShortenUrlRequest.java -> UrlShortenerRequest.java
- ShortenUrlResponse.java -> UrlShortenerResponse.java
- added @Validated to request.originUrl

healtheefart committed:
refactor : util관련 값을 설정파일로 관리

healtheefart committed:
refactor : request값 @Validated 추가

healtheefart committed:
refactor : UrlShortenerService.java 전면적 refactoring
- added @Transactional annotation
- createNewUrl()을 통해 새로 url 단축로직 분리

healtheefart committed:
refactor : BASEUtils의 필드값에 @Value 어노테이션을 통한 yml값 주입
- static field 라 바로 주입이 안 되어 un-static setter를 통해 값 주입

healtheefart committed:
feat : Error 관련 클래스 생성
- InvalidArgumentException : @Validated으로부터 던져진 exception
- NotFoundResourceException : 단축된 적 없는 shorten-url에 대한 redirect가 일어났을 때 던져진 exception
- index.html : error 발생 시 exception에 대한 message 출력

healtheefart committed:
refactor : changed css file

healtheefart committed:
feat : url 단축 요청에 대한 cnt 기능 추가

sunH0 commented on src/main/java/com/mission/ShortURLService/common/error/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,36 @@
+			.body(ErrorResponse.of(e));
+	}
+
+	@ExceptionHandler(MethodArgumentNotValidException.class)
```
검색해보니 MethodArgumentNotValidException 는 requestBody를 검증할 경우 나오는 예외라고 하네요!
PathVariable, RequestParam 에 대한 검증을 하고 싶다면 @Validated 어노테이션을 클래스 레벨의 Controller에 추가하면 되는 데 이때는 ConstraintViolationException  이 발생한다고 합니다!
자세한 이유는 좀 더 알아봐야 할 것 같네요!

healtheefart replied:
찾아보니 제가 @Validated를 사용했지만 @Valid으로 처리가 된 것 같아요!

> 컨트롤러의 메소드를 호출하는 과정에는 메소드의 값을 처리해주는 ArgumentResolver가 동작하는데, @Valid 역시 ArgumentResolver에 의해 처리가 된다.
대표적으로 @RequestBody가 있는 경우에는 Json 메세지를 객체로 변환해주는 작업이 ArgumentResolver의 구현체인  
RequestResponseBodyMethodProcessor가 처리하며, 이 내부에서 @Valid로 시작하는 어노테이션이 있을 경우에 유효성 검사를 진행한다. (이러한 이유로 @Valid가 아니라 커스톰 어노테이션인 @ValidMangKyu여도 동작한다.)
출처: https://mangkyu.tistory.com/174 [MangKyu's Diary:티스토리]

그리고 `@Valid`어노테이션은 controller단에서만 동작하고 만약 다른 계층에서 사용하고 싶다면 AOP로 동작하는 `@Validated`와 함께 사용해야한다고 해유
저는 `@Valid` 어노테이션을 처리해주는 `argumentResolver`에 대해서 좀 더 알아봐야할 거 같아요...!

sunH0 replied:
저도 이번에 Spring에 대해서 공부하면서 ArgumentResolver 와 MessageConverter 에 대해서 많이 알게 되었던 것 같아요! CS 공부용으로 알면 좋을 것 같습니다!
@Valid 와 @Validated는 기본적으로 비슷하지만 @Validated가 더 많은 기능을 갖고 있지 않을 까 싶네요! 같이 공부하고 공유하면 좋을 것 같습니다~

sunH0 commented on src/main/java/com/mission/ShortURLService/utils/BASE62Utils.java:
```diff
@@ -0,0 +1,42 @@
+		HEAD = head;
+	}
+	@Value("${shorten-algorithm-BASE62}")
+	public void setBASE62(String base62) {
```
굉장히 고민하신 것 같네요!
개인적인 생각은 빈으로 등록되는 클래스 내부에 정적 변수가 필요할까 의문이 들어요!
현재 두 필드가 전역 변수로 사용되지는 않아 보이고, @Component로 싱글톤으로 생성되니까 메모리 효율에도 의미가 있을 것 같지 않다는 생각이 드네요!
하정님은 어떻게 생각하시나요?
오히려 final 키워드가 좋을 지도...?

healtheefart replied:
흠... 왜 HEAD와 BASE62 필드가 static 메서드 내에서 전역변수로 사용되지 않아보이나요? 질문이 잘 이해가 가지 않아서요...!

sunH0 replied:
넵! 전역으로 사용되는 클래스라는 의견은 동의합니다!
하지만 @Component를 통해서 빈 객체와 싱글톤으로 사용 될 때와 비교해서  static(정적) 키워드를 지닌 효과가 있을지 고민되는 것 같아요! 저도 헷갈려서 이야기 나눠보면 좋을 것 같습니다!

sunH0 commented on src/main/resources/templates/index.html:
```diff
@@ -0,0 +1,52 @@
+      let response =
+          "<p>shorten url: <a href=" + $("#originUrl").val() + " target='_blank'>"
+          + parse.shortenUrl + "</a></p><br>"
+          + "<p>" + parse.cnt + " people shortened this url.</p>";
```
👍🏻👍🏻👍🏻

sunH0 commented on src/main/java/com/mission/ShortURLService/service/UrlShortenerService.java:
```diff
@@ -0,0 +1,46 @@
+	public UrlShortenerResponse getShortenUrl(UrlShortenerRequest request) {
+		Url url = urlShortenerRepository.findByOriginUrl(request.getOriginUrl())
+										.orElseGet(() -> createNewUrl(request));
+		url.increaseCnt();
```
지금 cnt 초기값이 1이라서 처음 url 생성하면 cnt가 바로 2가 될지도..??

healtheefart replied:
헉 그러게요...! 근데 제가 테스트 했을 때 새로 단축하는 url cnt가 1이었는데 왜 그렇게 됐을까요?😂
지금 확인해보니 increaseCnt query도 잘 돌아가는데 왜 1일까요??

sunH0 replied:
뭐죵?? 코드만 봤을 때는 이유를 모르겠네요 다음에 볼 때 같이 이유를 찾으면 좋을 것 같습니다! 너무 궁금하네요

sunH0 commented on src/main/java/com/mission/ShortURLService/dto/UrlShortenerRequest.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.validation.annotation.Validated;
+
+@Getter
+@Validated
```
이게 안붙으면 동작이 안될까요??

healtheefart replied:
아뇨ㅎㅎ 잘 되지요! 제가 왜 이걸 여기에 넣었을까요?😂

sunH0 commented on src/main/java/com/mission/ShortURLService/common/error/ErrorCode.java:
```diff
@@ -0,0 +1,17 @@
+
+@Getter
+public enum ErrorCode {
+	NOT_FOUND_SHORTEN_URL("U001", "존재하지 않는 URL 입니다."),
```
오 개별 code도 설정해 주셨군요!
저는 여기서 status code 도 관리해 주는 편입니다!

sunH0 commented on src/main/java/com/mission/ShortURLService/common/error/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,36 @@
+	@ExceptionHandler(NotFoundResourceException.class)
+	public ResponseEntity<ErrorResponse> NotFoundResourceExceptionHandler(NotFoundResourceException e) {
+		log.error("Exception: {}", e.getMessage());
+		return ResponseEntity
```
ResponseEntity 를 사용하셨네요! json 출력 형식이 궁금하네요!
중복되는 정보가 있을수도??

choiyoungkwon12 replied:
저도 이부분이 약간 궁금해지네요 🤔

healtheefart replied:
@choiyoungkwon12 잉?ㅋㅋㅋㅋ갑분영권님?!ㅋㅋㅋㅋㅋㅋ

healtheefart replied:
![image](https://user-images.githubusercontent.com/63666375/172201900-2c87fe4e-a8a7-422d-b9e5-4edf38b7e8be.png)
이렇게 나옵니다!

healtheefart replied:
중복되는 코드라면 ErrorResponse의 status 말씀...?

sunH0 replied:
넵 그 부분이 중복될까 궁금했어요 ResponseEntity.status(e.getHttpStatus())는 어디에 출력될까요??

healtheefart replied:
@sunH0 저 사진은 body만 보여지고,  status()로 선언한 status는 잘은 모르겠지만 헤더로 확인할 수 있을 것 같아요

sunH0 commented on src/main/java/com/mission/ShortURLService/common/error/exception/ErrorCodedException.java:
```diff
@@ -0,0 +1,25 @@
+	}
+
+	@Override
+	public String toString() {
```
이 toString을 언제 동작될까요??

healtheefart replied:
toString은 log의 getMessage() 대신 더 자세한 exception 내용을 확인하려 override 했었습니다!

sunH0 replied:
아하  e.getMessage() 를 사용하면 toString이 실행되군요! 하나 배워갑니다!

healtheefart replied:
앗 아니요! 지금은 toString으로 사용하진 않았지만 toString을 오버라이딩 해두어서 e.getMessage() 대신 e.toString()으로도 확인하려고 한 거에요!

sunH0 commented on src/main/java/com/mission/ShortURLService/common/error/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,36 @@
+		return ResponseEntity
+			.status(HttpStatus.BAD_REQUEST)
+			.body(ErrorResponse.of(
+				new InvalidArgumentException(INVALID_URL_FORMAT)
```
만약에 검증에서 2개 이상의 예외가 검출되면 어떻게 될까...??
저는 BindingResult를 사용해서 ErrorResponse에 정의 해 줍니다!

healtheefart replied:
오호 그런 게 있었군요...! 지금에서야 봤어요!! (commit 단위로 봤는데 지나쳤나봐요...!)
처음 듣는 개념이라 궁금한 게 있는데 
1. 예외가 여러 개 검출된다는 게 생소해서요...! 만약 검증할 필드를 여러 개 가진 request에 대한 검증이라면, exception 이 request의 모든 필드에 대한 검증을 마치고 throw 된다는 것이죠? 검증할 필드들을 순서대로 검증하다가 하나라도 unvalid하면 throw하는 것이 아니라...?!
2. [선호님의 ErrorResponse](https://github.com/study-for-Coding-Mission/Short-URL-Service/blob/sunH0%2Fmain/src/main/java/com/mission/shorturlservice/global/error/ErrorResponse.java) 에서 
33번째 줄 생성자가 궁금한데요..! errors의 초기화(?)없이 생성자의 파라미터 errors로 저렇게 바로 선언할 수 있나요? 뭔가 알고리즘문제를 풀 때 저렇게 하면 add와 같은 메서드를 사용하였을 때 에러가 났던 거 같아서요...!

sunH0 replied:
1번 에 대해서는 맞습니다~
2번에 대해서는 파라미터는 초기화 없이 어떻게 설정해도 오류가 있을 수 없다고 생각합니다!
혹시 질문에 대해 잘못 이해했다면 따로 질문해주시면 좋을 것 같아요!

sunH0 commented on gradlew.bat:
```diff
@@ -0,0 +1,89 @@
+@rem
```
.gitignore 가 설정이 안돼있는 것 같아요 ㅜㅜ

healtheefart replied:
.gitignore는 따로 제가 설정한 건 아니고 gitignore 자동완성 사이트를 통해서 설정한 건데요...!
gradlew.bat를 gitignore에 추가를 해야하는가에 대한 내용을 찾아보니 [이런 글](https://stackoverflow.com/questions/43954932/should-i-gitignore-my-gradlew) 이 있었습니다!
gradlew.bat 파일은 로컬에 gradle이 설치되지 않은 사람들이 해당 레포를 사용할 때 빠르게 그리고 알맞은 버전의 gradle을  다운받는 걸 도와주는 역할이라 굳이 git에 안 올릴 이유는 없다(?)라는 의견인 거 같아요!

sunH0 replied:
아하 저도 gitinore 적용에 대해서 참 골치가 아픕니다..... 같이 이야기 해보면 좋을 것 같아요!

sunH0 reviewed (commented):
리펙토링 하시느라 고생하셨어요~
코멘트 확인하시고 답변해주시면 감사하겠습니다!!

sunH0 commented on src/main/java/com/mission/ShortURLService/utils/BASE62Utils.java:
```diff
@@ -0,0 +1,42 @@
+
+	public static String encode(String value) {
+		StringBuilder sb = new StringBuilder(HEAD);
+		Integer intVal = Integer.valueOf(value);
```
갑자기 든 생각인데, 저희가 너무 짧은 url을 피하려고 DB id 초기값을 지정하려 했잖아요?
그냥 여기서 1000000 더해줘도 됐었네요 하하하

healtheefart replied:
오호라?!

choiyoungkwon12 commented on src/main/java/com/mission/ShortURLService/entity/Url.java:
```diff
@@ -0,0 +1,42 @@
+
+	@Id
+	@SequenceGenerator(name="seq_generator", sequenceName = "url_seq", initialValue = 100000)
+	@GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "seq_generator")
```
혹시 id 생성 전략을 왜 sequence로 하셨는지 알 수 있을까요??

healtheefart replied:
id를 encode(String 형태의 encoded-value)해서 `http://localhost:8080/redirect/{encoded-value}`이런 식으로 단축 url을 생성했는데요!
1부터 시작하는 id를 encode 해버리면 `http://localhost:8080/redirect/A` 이런 식으로 encoded-value의 길이가 너무 짧아 좀 더 긴 encoded-value를 주려고 고민하다가 sequence의 initialValue로 id 초기값을 좀 더 높게 주어 길이를 늘리려고 했어용
하지만 이 부분에 대해선 제 방법으로 해결하지 못한 부분들이 있어 아직 리팩토링할 것들이 많습니다 하핫

choiyoungkwon12 commented on src/main/java/com/mission/ShortURLService/common/error/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,36 @@
+			.body(ErrorResponse.of(
+				new InvalidArgumentException(INVALID_URL_FORMAT)
+			));
+	}
+}
```
만약 위에서 잡아주지 않는 exception이 발생한다면 어떻게 처리되나요??
Exception, RuntimeException으로 예상하지 못한 에러를 어떻게 처리할 것인지 생각해봐도 좋을 것 같습니다!

healtheefart replied:
앗...! 두 가지 exception에 대해서 코드를 짜고난 다음에 이 외의 Exception들은 @ExceptionHandler(Exception.class)로 잡아주려고 했는데 저 두 가지 exception 에만 너무 집중했는지 생각만 하고 추가를 안 해버렸군요...!

choiyoungkwon12 commented on src/main/java/com/mission/ShortURLService/entity/Url.java:
```diff
@@ -0,0 +1,42 @@
+	@Id
+	@SequenceGenerator(name="seq_generator", sequenceName = "url_seq", initialValue = 100000)
+	@GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "seq_generator")
+	private BigInteger id;
```
아이디는 BigInteger로 사용하신 이유를 알 수 있을까요?? 너무 궁금합니다 🤔

healtheefart replied:
initialValue를 높게 잡아 사용하다보니 범위가 무한대인 BigInteger를 사용했는데요! 위에서도 언급했지만 이 부분에 대해선 제 방법으로 해결하지 못한 부분들이 있어 아직 리팩토링할 것들이 많습니다 하핫

choiyoungkwon12 commented on src/main/java/com/mission/ShortURLService/common/error/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,36 @@
+
+	@ExceptionHandler(MethodArgumentNotValidException.class)
+	public ResponseEntity<ErrorResponse> InvalidArgumentExceptionHandler(MethodArgumentNotValidException e) {
+		log.error("Exception: {}", e.getMessage());
```
미션 구현 하면서는 크게 신경쓸 부분은 아니긴 한데요.
예전에 잭슨님 라이브세션때 하셨던 이야기가 생각납니다.
예상할 수 있는 예외에 대해서는 예외처리를 다해주기 때문에 로그레벨을 warn으로 하고 
예상하지 못한 에러들에 대해서는 error 레벨로 로그를 찍어서 error 로그가 찍히면 슬랙이든 어떤 형태든 모니터링 하면서 알람을 준다고 하셨던것 같습니다.
그래서 만약 신입이 와서 예외처리 로그 레벨을 다 error로 찍어 놓으면 깜짝 놀란다고....

나중에 로그 레벨을 어떤 것으로 찍을것인지를 생각해보는것도 좋을 것 같습니다! 👍

sunH0 replied:
기억이 나네요! 저도 로깅에 대해서 다시 고민해야 할 것 같아요!
취업과 상관없이 꾸준히 하고자 하는 스터디라 영권님 생각있으시면 말해주세요 ㅎㅎ

healtheefart replied:
오호 현업에선 그럴 수 있겠군요! 근데 이 얘기를 들으니 괜히 더 error로 로그를 찍고 싶은 이유는 뭘까요...?🤔

choiyoungkwon12 commented on src/main/java/com/mission/ShortURLService/repository/UrlShortenerRepository.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
`@Repository` 어노테이션을 붙여준 이유가 있을까요??

평소에 저는 JpaRespository extends하면 그냥 안붙여도 돼서 생각없이 안붙였었는데 `@Repository`를 사용하셔서 어떻게 JpaRespository를 상속하면 레포지토리로 인식하고 사용이 될까??  궁금해지네요

healtheefart replied:
그냥 사용할 respository라 @Repository 를 붙여두었는데 보통 붙이지 않는군요..!
음 근데 이 리뷰를 보고 @Repository 들을 까보니 궁금해진 것이 있는데요
해당 레포가 확장한 인터페이스(JpaRepository)에는 @NoRepositoryBean이라고 붙여져있는데 그럼 레포지토리는 빈으로 등록해서 사용하지 않나요?

sunH0 replied:
오! 저도 궁금해서 찾아봤는데
https://velog.io/@max9106/wtc-learning11
이 글을 읽으시면 좋을 것 같아요!

choiyoungkwon12 reviewed (commented):
하정님, 선호님 두분이서 이런걸 하고 계셨군요! 재밌어 보이네요 ㅎㅎ 
다 읽진 못했지만 pr에 서로 코드리뷰 하는 부분과 작성하신 코드들 구경하면서 많이 배웠습니다!! 그리고 염탐하다가 궁금한 점 남겼습니다! 
👍 👍 👍

healtheefart committed:
refactor : BASE62Utils 리팩토링
- id Integer로 변경
- encoded value를 6자 이상으로 반환하기 위해 BASE62Utils에서 BigInteger 연산 사용

healtheefart committed:
refactor : 정의되지 않은 exception 처리 및 log 레벨 warn으로 변경

healtheefart committed:
chore : requestDto에서 @Validated 제거, repository에서 @Repository 제거
